src/mlpack/core/data/load_image.hpp

```diff
@@ -0,0 +1,212 @@
+
+#include <mlpack/core.hpp>
+#include <mlpack/core/util/log.hpp>
+#include <mlpack/prereqs.hpp>
+
+#if defined(__cplusplus) && __cplusplus >= 201703L && defined(__has_include)
+    && __has_include(<filesystem>)
+#include <filesystem>
+#define HAS_FILESYSTEM 1
+namespace fs = std::filesystem;
+#endif
+
+#include "extension.hpp"
+
+#ifndef STB_IMAGE_IMPLEMENTATION
+#define STB_IMAGE_IMPLEMENTATION
+#include <stb_image.h>
+#endif
+
+#ifndef STB_IMAGE_WRITE_IMPLEMENTATION
+#define STB_IMAGE_WRITE_IMPLEMENTATION
+#include <stb_image_write.h>
+#endif
+
+namespace mlpack {
+namespace data {
+
+/**
+ * Loads a matrix with image. It also supports loading image from 
+ * an entire directory.
+ * @code
+ * arma::Mat<unsigned char> img;
+ * data::Image loader;
+ * loader.Load("test_image.png", std::move(img));
+ * @endcode
+ *
+ * The supported types of files are:
+ * 
+ * - JPEG baseline & progressive (12 bpc/arithmetic not supported,
+ *    same as stock IJG lib)
+ * - PNG 1/2/4/8/16-bit-per-channel
+ * - TGA (not sure what subset, if a subset)
+ * - BMP non-1bpp, non-RLE
+ * - PSD (composited view only, no extra channels, 8/16 bit-per-channel)
+ * - GIF (*comp always reports as 4-channel)
+ * - HDR (radiance rgbE format)
+ * - PIC (Softimage PIC)
+ * - PNM (PPM and PGM binary only).
+ */
+class Image
+{
+ public:
+  /**
+   * Image default constructor.
+   */
+  Image();
+
+  /**
+   * Instantiate the Image object with the image width, height, channels.
+   *
+   * @param width Matrix width for the output matrix.
+   * @param height Matrix height for the output matrix.
+   * @param channels Matrix channels for the output matrix.
+   */
+  Image(const size_t width, const size_t height, const size_t channels);
+
+  /**
+   * Checks if the given image filename is supported.
+   *
+   * @param filename Name of the image file.
+   * @return Boolean value indicating success if it is an image.
+   */
+  bool ImageFormatSupported(const std::string& fileName, bool save = false);
+
+  /**
+   * Load the image file into the given matrix.
+   *
+   * @param fileName Name of the image file.
+   * @param flipVertical Flip the image vertical upon loading.
+   * @param outputMatrix Matrix to load into.
+   * @return Boolean value indicating success or failure of load.
+   */
+  bool Load(const std::string& fileName,
+            bool flipVertical,
+            arma::Mat<unsigned char>&& outputMatrix);
+
+  /**
+   * Load the image file into the given matrix.
+   *
+   * @param fileName Name of the image file.
+   * @param flipVertical Flip the image vertical upon loading.
+   * @param width Width of the image file.
+   * @param height Height of the image file.
+   * @param channels Channels of the image file.
+   * @param outputMatrix Matrix to load into.
+   * @return Boolean value indicating success or failure of load.
+   */
+  bool Load(const std::string& fileName,
+            bool flipVertical,
+            arma::Mat<unsigned char>&& outputMatrix,
```

Comment:
Do you think we should go with a simple reference for every output? Might look cleaner to do:

```c++
arma::mat image
Load("test.png", false, image, ...);
```

also I was thinking it might be useful to put the ``flipVertical`` at the end, don't think this option is something that a user should pass each time.

Reply:
Using simple reference may require typecasting at various stages. like when I am using `inputmartix.col(i)` it needs to be typecasted to `arma::Mat<uchar>` for every col. Hence, I did not use a simple reference. However, if you want I can add a simple reference.

Reply:
Not sure I get the example, can you provide an actual example where you have to cast if you change the interface from ``arma::Mat<unsigned char>&& outputMatrix,`` to ``arma::Mat<unsigned char>& outputMatrix,``? I looked over the code and I think if you just use a reference and remove the move call it should be just fine.

Reply:
Here https://github.com/mlpack/mlpack/blob/4eae4d4d8faf6a450fd2a78f73f78adc92e9b62d/src/mlpack/core/data/load_image_impl.hpp#L194
This is of type subview and needs to be converted to Mat, else the compiler throws an error.

Reply:
Now I see what you mean, what about we do:

```c++
arma::Mat<unsigned char>(outputMatrix.colptr(i), outputMatrix.n_rows, 1, false, true)
```